src/util/positioning.ts

```diff
@@ -1,167 +1,49 @@
-
-    if (this.getStyle(element, 'position') === 'fixed') {
-      elPosition = element.getBoundingClientRect();
-      elPosition = {
-        top: elPosition.top,
-        bottom: elPosition.bottom,
-        left: elPosition.left,
-        right: elPosition.right,
-        height: elPosition.height,
-        width: elPosition.width
-      };
-    } else {
-      const offsetParentEl = this.offsetParent(element);
-
-      elPosition = this.offset(element, false);
-
-      if (offsetParentEl !== document.documentElement) {
-        parentOffset = this.offset(offsetParentEl, false);
-      }
-
-      parentOffset.top += offsetParentEl.clientTop;
-      parentOffset.left += offsetParentEl.clientLeft;
-    }
-
-    elPosition.top -= parentOffset.top;
-    elPosition.bottom -= parentOffset.top;
-    elPosition.left -= parentOffset.left;
-    elPosition.right -= parentOffset.left;
-
-    if (round) {
-      elPosition.top = Math.round(elPosition.top);
-      elPosition.bottom = Math.round(elPosition.bottom);
-      elPosition.left = Math.round(elPosition.left);
-      elPosition.right = Math.round(elPosition.right);
-    }
-
-    return elPosition;
-  }
+const placementSeparator = /\s+/;
+const spacesRegExp = /  +/gi;
+
+const startPrimaryPlacement = /^start/;
+const endPrimaryPlacement = /^end/;
+const startSecondaryPlacement = /-top$/;
+const endSecondaryPlacement = /-bottom$/;
+function convertPlacementToPopper(placement: Placement): PopperPlacement {
+  const newPlacement = placement.replace(startPrimaryPlacement, 'left')
+                           .replace(endPrimaryPlacement, 'right')
+                           .replace(startSecondaryPlacement, '-start')
+                           .replace(endSecondaryPlacement, '-end') as PopperPlacement;
+  return newPlacement;
+}
 
-  offset(element: HTMLElement, round = true): ClientRect {
-    const elBcr = element.getBoundingClientRect();
-    const viewportOffset = {
-      top: window.pageYOffset - document.documentElement.clientTop,
-      left: window.pageXOffset - document.documentElement.clientLeft
-    };
-
-    let elOffset = {
-      height: elBcr.height || element.offsetHeight,
-      width: elBcr.width || element.offsetWidth,
-      top: elBcr.top + viewportOffset.top,
-      bottom: elBcr.bottom + viewportOffset.top,
-      left: elBcr.left + viewportOffset.left,
-      right: elBcr.right + viewportOffset.left
-    };
-
-    if (round) {
-      elOffset.height = Math.round(elOffset.height);
-      elOffset.width = Math.round(elOffset.width);
-      elOffset.top = Math.round(elOffset.top);
-      elOffset.bottom = Math.round(elOffset.bottom);
-      elOffset.left = Math.round(elOffset.left);
-      elOffset.right = Math.round(elOffset.right);
-    }
-
-    return elOffset;
+const popperStartPrimaryPlacement = /^left/;
+const popperEndPrimaryPlacement = /^right/;
+function convertPlacementToBsClasses(baseClass: string, placement: PopperPlacement): string {
```

Comment:
maybe something  like `getBootstrapBaseClassPlacement`. I was struggling to understand what is does.

Also unit tests for these kind of utility functions would be nice...